### ISSUE TITLE
Events stop

### ISSUE THREAD

**Issue by sabey** [dangrie158/dolce#17]
Hey, sorry to bother you again but I seem to have found an issue with stale events.

I'm running Dolce using `docker-socket-proxy` `DOCKER_HOST: host:2375 DOCKER_TRANSPORT: tcp` and this works initially, however the longer the container runs for at some point the events stop appearing.

Is it possible that the tcp connection to the proxy is breaking and not reconnecting?

Restarting the container will not pick up the missed events either, so that would leave a temporary solution of restarting the container on a cron out of the question.

Thanks!

**Comment by dangrie158** [comment]
Hi,
what you are probably seeing is the exponential event backoff. In its standard Configuration, dolce is meant to be used to monitor containers that should always be running and immediately notify you if any container stops. However, sometimes a problem occurs where the container restarts in a loop, thus sending a never ending stream of events. To mitigate this issue, dolce will exponentially increase the time between notifications, so on the first event you get notified immediately, then if any more events occur the earliest date for the next notification is 10s, then 100s after that, then 1000 and so on until a configurable maximum time is reached. In between notifications, the events are collected and scheduled for resubmission after the backoff timeout, even spanning a restart.

If you don't want this behaviour and are fine getting pinged on every event, see issue #15 and the corresponding [documentation page](https://dangrie158.github.io/dolce/master/advanced/notification-backoff/) to find out how to disable this behaviour.

Let me know if this is indeed the issue, as it would explain the behaviour you experience.

**Comment by sabey** [comment]
Hey, thanks for the reply. My problem is that I'm not seeing any events (not to mention notifications) in the log at all.

I am currently using those environment variables but they are set to this (I'll try 0 just to be safe):
```
DOLCE_MIN_TIMEOUT: 10
DOLCE_MAX_TIMEOUT: 3600
DOLCE_MULTIPLIER: 10
``

These are my current logs:
```
2024-03-20T20:56:05.288876935Z INFO	main	 starting dolce container monitor v2.8.0
2024-03-20T20:56:05.296843517Z INFO	main	 created lockfile /var/run/dolce/lockfile for pid 7
2024-03-20T20:56:05.328791043Z INFO	main	 connected to Docker 24.0.7 (API: 1.43)
2024-03-20T20:56:05.328824149Z INFO	main	 supervision mode set to ALL
2024-03-20T20:56:05.329045315Z INFO	notifier	 creating AppriseNotifier
2024-03-20T20:56:05.331975520Z INFO	main	 no delivery scheduled. waiting for events
2024-03-20T20:56:05.339244106Z INFO	main	 new container event received: <"dolce": start>
2024-03-20T20:56:06.443370049Z INFO	main	 new container event received: <"dolce": restart>
2024-03-20T20:56:15.522363174Z INFO	main	 sending events notification to all registered notifiers with 2 events
2024-03-20T20:56:16.362998268Z INFO	main	 earliest next notification at 3/20/2024, 8:57:55 PM
```
I've since had a few different docker container be restarted by watchtower during this time so my assumption is that I should have seen the events in the logs since then? This current instance of Dolce has been running for 12 hours since the initial 2 dolce start/restart events, but nothing else has appeared.

I've manually restarted a few containers (on this current instance) just now hoping to see something in the logs but nothing happens. I do know manually restarting containers/watchtower will trigger the event logs on a newly started Dolce instance which is why I suspected something has happened to the tcp socket.

**Comment by sabey** [comment]
Restarted Dolce with the 0 values and waited an hour before restarting a random service... saw no events in the logs or a notification.

**Comment by dangrie158** [comment]
Mhh, this is indeed strange. You are right in that you should definitely see the incoming events in the logs immediately.

I never had this issue, neither with nor without the proxy. Would it be possible for you to test if the error also occurs when connecting to the docker socket directly? 

Can you also in parallel curl the events on both the socket and the proxy like so:

`curl --unix-socket /var/run/docker.sock http://localhost/v1.27/events` for the socket and 
`curl  http://<address to the proxy>/v1.27/events` for the proxy connection. Maybe the proxy times out on the long polling endpoint but I never observed this before.

**Comment by sabey** [comment]
I tested two different Dolce instances on two different servers, one using `docker.sock` and the other using `docker-socket-proxy`.
I ran two curls on each server, one to the `docker.sock` and the other to `docker-socket-proxy` and left it running for a few hours without any disconnects, the output was even the same.
After the wait I then tested the event system by restarting a container on both servers. All of the curls showed the events appear, but only the Dolce instance using `docker.sock` showed an event appear.

**Comment by dangrie158** [comment]
Thank you very much for testing, this helps immensely narrowing down the problem. 
Since the Problem only occurs in the proxy configuration, I suspect that the proxy kills the connection after a certain amount of time. However This should cause the dolce container to quit and then, depending on your configuration, restart again. At least something should appear in the logs...

I just discovered: https://github.com/Tecnativa/docker-socket-proxy/issues/21 which seems no describe the problems your having pretty well. Still, I don't know why this does not kill the dolce process as the last thing that happens in the `main.ts` file is iterating over all events until that steam is closed, so by that time the process should stop.

I'll try to investigate, maybe it's just as simple as reconnecting once the steam times out 🤷‍♂️

**Comment by dangrie158** [comment]
Good news,
I was able to reproduce the problem. Indeed it seems like the docker-proxy closes the connection after a while. The reason why this never came up during testing is because it only happens when there are no events and thus the proxy thinks the connection timed out. I don't use the proxy for my installations so I never noticed.

I have to test a little bit more and would then push a fix to the ,master branch that you can then checkout. If the problem is resolved for you as well, I will release a new minor version.

**Comment by dangrie158** [comment]
OK, seems like I was able to fix the issue.

I just pushed the necessary changes.
If your schedule allows it, could you checkout the `:master` tag of the image on docker hub and report back if this fixes the problem?

Thank you so much for reporting the issue and helping to narrow down the root cause!

**Comment by dangrie158** [comment]
Ah, never mind. I will need to make another change tomorrow, seems like the last event gets sent over and over again after the reconnection.

I will keep you posted! 😊

**Comment by dangrie158** [comment]
All right! I just pushed the new changes. Everything looks good now for me. Let me know if you get around to test the current `master` tag on docker hub to make sure the issue is fixed

**Comment by sabey** [comment]
Everything seems to work perfectly now :)

The logs seem to line up with this... looks like there's a haproxy with a 10m timeout:
https://github.com/Tecnativa/docker-socket-proxy/blob/master/haproxy.cfg

I appreciate the time you took to fix this, thanks again!
```
2024-03-23T21:58:31.418812394Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T22:08:36.428443005Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T22:18:41.433487793Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T22:28:46.439350777Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T22:38:51.446427958Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T22:48:56.451770612Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T22:59:01.457410938Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T23:09:06.463674790Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T23:19:11.469404217Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
2024-03-23T23:29:16.474891747Z INFO	main	 event stream closed unexpectedly, reconnecting in 5 seconds
```

**Comment by dangrie158** [comment]
Nice! Thanks for confirming the fix!

I just released [v2.9.0](https://github.com/dangrie158/dolce/tree/v2.9.0) containing all the changes.

>  with a 10m timeout:

Interestingly I had to wait up to 22 min while debugging the issue before the timeout occurred 🤷‍♂️

> reconnecting in 5 seconds

FYI: when dolce restarts it queries the docker socket for all events that happened when the disconnect occurred, so you won't loose any events during this 5 second timeout
